src/Catalyst.Core.Modules.Web3/Web3EthApiExtensions.cs

```diff
@@ -0,0 +1,148 @@
+            {
+                throw new Exception($"Delta not found '{cid}'");
+            }
+
+            return new DeltaWithCid
+            {
+                Delta = delta,
+                Cid = cid
+            };
+        }
+
+        public static DeltaWithCid GetDeltaWithCid(this IWeb3EthApi api, BlockParameter block)
+        {
+            Cid cid;
+            switch (block.Type)
+            {
+                case BlockParameterType.Earliest:
+                    cid = api.DeltaCache.GenesisHash;
+                    break;
+                case BlockParameterType.Latest:
+                    cid = api.DeltaResolver.LatestDelta;
+                    break;
+                case BlockParameterType.Pending:
+                    cid = api.DeltaResolver.LatestDelta;
+                    break;
+                case BlockParameterType.BlockNumber:
+                    var blockNumber = block.BlockNumber.Value;
+                    cid = api.DeltaResolver.Resolve(blockNumber);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            return api.GetDeltaWithCid(cid);
+        }
+
+        public static PublicEntry ToPublicEntry(this IWeb3EthApi api, TransactionForRpc transactionCall, Keccak root)
+        {
+            return new PublicEntry
+            {
+                Nonce = (ulong) api.StateReader.GetNonce(root, transactionCall.From),
+                SenderAddress = transactionCall.From.Bytes.ToByteString(),
+                ReceiverAddress = transactionCall.To?.Bytes.ToByteString() ?? ByteString.Empty,
+                GasLimit = (ulong) transactionCall.Gas.GetValueOrDefault(),
+                GasPrice = transactionCall.GasPrice.GetValueOrDefault().ToUint256ByteString(),
+                Amount = transactionCall.Value.GetValueOrDefault().ToUint256ByteString(), 
+                Data = transactionCall.Data?.ToByteString() ?? ByteString.Empty
+            };
+        }
+
+        public static CallOutputTracer CallAndRestore(this IWeb3EthApi api, TransactionForRpc transactionCall, DeltaWithCid deltaWithCid)
+        {
+            var parentDelta = deltaWithCid.Delta;
+            Keccak root = parentDelta.StateRootAsKeccak();
+
+            if (transactionCall.Gas == null)
+            {
+                transactionCall.Gas = parentDelta.GasLimit;
+            }
+
+            var publicEntry = api.ToPublicEntry(transactionCall, root);
+
+            var newDelta = deltaWithCid.CreateOneOffDelta(publicEntry);
+
+            CallOutputTracer callOutputTracer = new CallOutputTracer();
+
+            api.StateProvider.StateRoot = root;
+            api.Executor.CallAndRestore(newDelta, callOutputTracer);
+            api.StateProvider.Reset();
+            api.StorageProvider.Reset();
+            return callOutputTracer;
+        }
+    }
+
+    public struct DeltaWithCid
+    {
+        public Delta Delta;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4585091)